File: 1_GuidedExercise/idea_11.rb
class Idea11
  # Can you use extract and introduce variable to make this code clearer?
  def under_abstracted(operation, first_argument, second_argument=nil)
    if second_argument != nil
      number_list = [first_argument, second_argument]
    else
      number_list = first_argument
    end

    if number_list == []
      return 0
    end

    if operation == '+'
      symbol = :+
      result = number_list.shift
      result = number_list.reduce(result, symbol)
    elsif operation == 'x'
      symbol = :*
      result = number_list.shift
      result = number_list.reduce(result, symbol)
    elsif operation == '-'
      symbol = :-
      result = number_list.shift
      result = number_list.reduce(result, symbol)
    elsif operation == '÷'
      symbol = :/
      result = number_list.shift
      result = number_list.reduce(result, symbol)
    end

    result
  end
end

File: js/1_GuidedExercise/src/idea_11.js
// Task: Use extract method, introduce variable and introduce parameter to make this code clearer.
const underAbstracted = (operation, firstArgument, secondArgument = null) => {
    let numberList
    if (secondArgument !== null) {
        numberList = [firstArgument, secondArgument]
    } else {
        numberList = firstArgument
    }

    if (numberList.length === 0) {
        return 0
    }

    let result
    if (operation === '+') {
        const calc = (a, b) => a + b
        result = numberList.shift()
        result = numberList.reduce(calc, result)
    } else if (operation === 'x') {
        const calc = (a, b) => a * b
        result = numberList.shift()
        result = numberList.reduce(calc, result)
    } else if (operation === '-') {
        const calc = (a, b) => a - b
        result = numberList.shift()
        result = numberList.reduce(calc, result)
    } else if (operation === '÷') {
        const calc = (a, b) => a / b
        result = numberList.shift()
        result = numberList.reduce(calc, result)
    }

    return result
}

module.exports = {underAbstracted}